install/install.sh

```diff
@@ -50,55 +50,163 @@ release_name() {
-for shell_profile in "${HOME}/.bashrc" "${HOME}/.zshrc"
-do
-    if ! grep "${path_modification}" "${shell_profile}" &>/dev/null; then
-        verbose "Updating \$PATH in ${shell_profile} to include ${latest_directory}"
-        printf "\n${path_modification}\n" >> $shell_profile
+    verbose "Creating ${azureauth_directory}"
+    mkdir -p $azureauth_directory
+
+    verbose "Downloading ${release_url} to ${tarball}"
+    if ! curl -sfL $release_url > $tarball; then
+        error "Failed to download azureauth ${version}"
+        exit 1
+    fi
+
+    if [ -d "${target_directory}" ]; then
+        verbose "Removing pre-existing target directory at ${target_directory}"
+        rm -rf $target_directory
     fi
-done
 
-echo "Installed azureauth ${version}!"
+    verbose "Extracting ${tarball} to ${target_directory}"
+    mkdir -p $target_directory
+    tar -xf $tarball -C $target_directory
+
+    # The files extracted from the tarball are all executable, but only two need to be.
+    chmod -x ${target_directory}/*
+    chmod +x ${target_directory}/azureauth
+
+    verbose "Removing ${tarball}"
+    rm $tarball
+
+    verbose "Linking ${latest_directory} to ${target_directory}"
+    # It's very important we use -n and -f here or the symlink won't actually be overwritten during upgrades.
+    ln -snf $target_directory $latest_directory
+
+    # We currently only support automatically appending $PATH modifications for the default
+    # Bash and ZSH profiles as the syntax is identical.
+    path_modification='export PATH="${PATH}:${HOME}/.azureauth/latest"'
+    for shell_profile in "${HOME}/.bashrc" "${HOME}/.zshrc"
+    do
+        if ! grep "${path_modification}" "${shell_profile}" &>/dev/null; then
+            verbose "Updating \$PATH in ${shell_profile} to include ${latest_directory}"
+            printf "\n${path_modification}\n" >> $shell_profile
+        fi
+    done
+
+    echo "Installed azureauth ${version}!"
+}
+
+install_post_0_4_0() {
+    : ${AZUREAUTH_REPO='AzureAD/microsoft-authentication-cli'}
+    repo="${AZUREAUTH_REPO}"
+    release_file="$(release_name).tar.gz"
+    release_url="https://github.com/${repo}/releases/download/${version}/${release_file}"
+
+    : ${AZUREAUTH_INSTALL_DIRECTORY="${HOME}/.azureauth"}
+    azureauth_directory="${AZUREAUTH_INSTALL_DIRECTORY}"
+    target_directory="${azureauth_directory}/${version}"
+    tarball="${azureauth_directory}/${release_file}"
+
+    # Ignore profile updates if the user has requested we not touch their profile(s).
+    : ${AZUREAUTH_NO_UPDATE_PATH=""}
+    no_update_path="${AZUREAUTH_NO_UPDATE_PATH}"
+
+    verbose "Installing using post-0.4.0 method"
+
+    verbose "Creating ${azureauth_directory}"
+    mkdir -p $azureauth_directory
+
+    verbose "Downloading ${release_url} to ${tarball}"
+    if ! curl -sfL $release_url > $tarball; then
+        error "Failed to download azureauth ${version}"
+        exit 1
+    fi
```

Comment:
What's the usage of this in PowerShell? I did search before adding this comment and did not find anything substantial.

Reply:
This is `/bin/sh` syntax, not PowerShell. In shells like `/bin/sh` (Bash and ZSH count here too) the `fi` statement closes a corresponding `if` (note that it's just `if` backwards!).

Reply:
I'm **not** a fan of this syntax. It's one of my least favorite parts of shell scripting.

Reply:
Fun stuff! Learnt something.